### PR TITLE
Adds linalg.det alias, fixes outer alias, updates alias testing

### PR DESCRIPTION
**Pull request by mruberry**
This PR:

- updates test_op_normalization.py, which verifies that aliases are correctly translated in the JIT
- adds torch.linalg.det as an alias for torch.det
- moves the torch.linalg.outer alias to torch.outer (to be consistent with NumPy)

The torch.linalg.outer alias was put the linalg namespace erroneously as a placeholder since it's a "linear algebra op" according to NumPy but is actually still in the main NumPy namespace. 

The updates to test_op_normalization are necessary. Previously it was using method_tests to generate tests, and method_tests assumes test suites using it also use the device generic framework, which test_op_normalization did not. For example, some ops require decorators like `skipCPUIfNoLapack`, which only works in device generic test classes. Moving test_op_normalization to the device generic framework also lets these tests run on CPU and CUDA.

Continued reliance on method_tests() is excessive since the test suite is only interested in testing aliasing, and a simpler and more readable `AliasInfo` class is used for the required information. An example impedance mismatch between method_tests and the new tests, for example, was how to handle ops in namespaces like torch.linalg.det. In the future this information will likely be folded into a common 'OpInfo' registry in the test suite. 

The actual tests performed are similar to what they were previously: a scripted and traced version of the op is run and the test verifies that both graphs do not contain the alias name and do contain the aliased name.

The guidance for adding an alias has been updated accordingly. 

cc @mattip 

Note:

@ngimel suggests:
- deprecating and then removing the `torch.ger` name
- reviewing the implementation of `torch.outer`

### CODE REVIEW

File: test/test_op_normalization.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    for info in alias_infos:
+
+        @torch.no_grad()
+        def _test(self, device, info=info):
+            tensor = torch.tensor
+            op = info.alias_op
+            is_inplace = info.alias_name.endswith('_')
+
+            # NOTE: this workaround necessary to satisfy the JIT, which
+            #   does allow Python aliasing or direct calling of torch.Tensor
```
Comment (mruberry) [inline test/test_op_normalization.py:62]: doesn't

File: test/test_op_normalization.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            if is_inplace:
+                fn_template = '''
+                    def _fn(t):
+                        return t.{alias_name}({args})
+                '''
+                arg_string = ', '.join((str(arg) for arg in info.args))
+                script = fn_template.format(alias_name=info.alias_name, args=arg_string)
+            else:
+                # Checks that scripting converts aliases
+                # NOTE: scripting doesn't support splatting args, so this
```
Comment (mruberry) [inline test/test_op_normalization.py:73]: Move this note up since splatting happens in both cases

File: test/test_op_normalization.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            # Compiles script
+            scripted = torch.jit.CompilationUnit(script)._fn
+
+            # Acquires and checks the graph remaps the alias
+            scripted(info.input)
+            graph = scripted.graph_for(info.input)
+            FileCheck().check(info.original_name).check_not(info.alias_name).run(graph)
+
+            # Checks that tracing converts aliases
+            # NOTE: tracing has no problem splatting args
```
Comment (mruberry) [inline test/test_op_normalization.py:91]: also can call torch.Tensor.<op>

File: test/test_jit.py [hunk 4]
```diff
@@ -15166,14 +15165,6 @@ def parameter_script(x: torch.nn.Parameter):
     'test_nn_max_pool1d_with_indices',
 }
 
-def check_alias_annotation(method_name, args, kwargs):
-    formals, tensors, actuals = get_script_args(args)
-    call = get_call(method_name, 'method', actuals, kwargs)
-    script = script_template.format(', '.join(formals), call)
-    CU = torch.jit.CompilationUnit(script)
-    torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), method_name)
```
Comment (eellison) [inline test/test_jit.py:15174]: this is no longer being called - it's not checking op normalization of aliasing, it's checking alias annotations of schemas

File: test/test_linalg.py [hunk 5]
```diff
@@ -4,33 +4,65 @@
...
 
-        def fn(a, b):
-            return torch.linalg.outer(a, b)
-        scripted_outer = torch.jit.script(fn)
+        def _fn(a, b):
+            return torch.outer(a, b)
+
+        ops = (torch.ger, torch.Tensor.ger,
+               torch.outer, torch.Tensor.outer,
+               torch.jit.script(_fn))
```
Comment (mruberry) [inline test/test_linalg.py:28]: Remove jit variants (for now), maybe add back systematically later

File: test/test_linalg.py [hunk 6]
```diff
@@ -4,33 +4,65 @@
...
+            torch.randn((4, 2, 26, 26), device=device, dtype=dtype))
+
+        def _fn(t):
+            return torch.linalg.det(t)
 
-        self.assertEqual(actual, expected, exact_device=False)
-        self.assertEqual(actual, original)
-        self.assertEqual(actual, scripted)
+        ops = (torch.det, torch.Tensor.det,
+               torch.linalg.det, torch.jit.script(_fn))
```
Comment (mruberry) [inline test/test_linalg.py:51]: Remove jit variant (see above)

File: test/test_linalg.py [hunk 7]
```diff
@@ -4,33 +4,65 @@
...
+        for t in tensors:
+            expected = np.linalg.det(t.cpu().numpy())
+            for op in ops:
+                actual = op(t)
+                self.assertEqual(actual, expected)
 
+        # NOTE: det requires a 2D+ tensor
+        t = torch.randn(1, device=device, dtype=dtype)
+        for op in ops:
```
Comment (mruberry) [inline test/test_linalg.py:60]: update to self.assertRaises since no longer testing jit

File: test/test_jit.py (hunk 4 above)
Comment (mruberry) [reply test/test_jit.py:15174]: Aha! I wanted to check on that. I'll revert this change.

File: test/test_op_normalization.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                 decorators=()):  # decorators to apply to the test
+        self.alias_name = alias_name
+        self.alias_op = alias_op
+        self.original_name = original_name
+        self.input = input
+        self.args = args
+        self.decorators = decorators
+
+alias_infos = (
+    AliasInfo('absolute', torch.absolute, 'abs',
```
Comment (eellison) [inline test/test_op_normalization.py:32]: I think it's kind of nice to reuse the op registry instead of creating a new one...

**Review by eellison — APPROVED** [review]
op alias changes LGTM

File: test/test_op_normalization.py (hunk 8 above)
Comment (mruberry) [reply test/test_op_normalization.py:32]: It is! But check this out: https://github.com/pytorch/pytorch/pull/41662.

So now we can update this to use the fancy new op registry.
